File: src/types/rule/FocusRule.ts
import * as DB from "tsondb/schema/dsl"
import { SubjectIdentifier } from "../_Identifier.js"
import { RulePrerequisites } from "../_Prerequisite.js"
import { NestedTranslationMap } from "../Locale.js"
import { Errata } from "../source/_Erratum.js"
import { src } from "../source/_PublicationRef.js"

export const FocusRule = DB.Entity(import.meta.url, {
  name: "FocusRule",
  namePlural: "FocusRules",
  type: () =>
    DB.Object({
      subject: DB.Required({
        comment: "The associated subject.",
        type: SubjectIdentifier(),
      }),
      level: DB.Required({
        comment: "The focus rule’s level.",
        type: DB.Integer({ minimum: 1, maximum: 4 }),
      }),
      isMissingImplementation: DB.Required({
        comment:
          "Has the focus rule not been implemented in Optolith yet? This is also true if the focus rule does not (currently) apply to any Optolith feature.",
        type: DB.Boolean(),
      }),
      prerequisites: DB.Optional({
        type: DB.IncludeIdentifier(RulePrerequisites),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "FocusRule",
        DB.Object({
          name: DB.Required({
            comment: "The focus rule’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          description: DB.Required({
            comment: "The description of the focus rule.",
            type: DB.String({ minLength: 1, isMarkdown: true }),
          }),
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

File: Sources/OptolithDatabaseSchema/Types/Rule/FocusRule.swift
import FileDB

@Model
public struct FocusRule {

  /// The associated subject.
  @Relationship(Subject.self)
  let subject: Subject.ID

  /// The focus rule’s level.
  @Minimum(1)
  @Maximum(4)
  let level: Int

  /// Has the focus rule not been implemented in Optolith yet? This is also true if the focus rule does not (currently) apply to any Optolith feature.
  let isMissingImplementation: Bool

    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation { // FocusRuleTranslation

        /// The focus rule’s name.
        @MinLength(1)
        let name: String

        /// The description of the focus rule.
        @MinLength(1)
        let description: String

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}
